keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,757 @@
++ _driverless_ - a use-case where no driver exists to support the backend object store. Some COSI automation may still be provided, but backend operations such as creating a bucket or minting credentials will not occur.
++ _greenfield bucket_ - a new backend bucket created by COSI.
++ _green-to-browfield_ - a use-case where COSI creates a new bucket on behalf of a user, and then supports ways for other users in the cluster to share this bucket.
++ _provisioner_ - a generic term meant to describe the combination of a sidecar and driver. "Provisioning" a bucket can mean creating a new bucket or granting access to an existing bucket.
++ _sidecar_ - a Kubernetes-aware container (usually paired with a driver) that is responsible for fullfilling COSI requests with the driver via gRPC calls. The sidecar's access can be restricted to the namespace where it resides, which is expected to be the same namespace as the provisioner. The COSI sidecar is provided by the Kubernetes community.
++ _static provisioning_ - custom resource creation is done by the admin rather than via COSI automation. This may also include _driverless_ but they are independent.
+
+# Proposal
+
+## APIs
+
+### Storage APIs
+
+#### BucketRequest
+
+A user facing, namespaced custom resource requesting provisioning of a new bucket, or requesting access to an existing bucket. A `BucketRequest` (BR) lives in the app's namespace.  In addition to a `BucketRequest`, a [BucketAccessRequest](#bucketaccessrequest) is necessary in order to grant credentials to access the bucket. BRs are required for both greenfield and brownfield uses.
+
+There is a 1:1 mapping of a `BucketRequest` and the cluster scoped `Bucket` instance, meaning there is a single `Bucket` instance for every BR.
+
+```
+apiVersion: cosi.io/v1alpha1
+kind: BucketRequest
+metadata:
+  name:
+  namespace:
+  labels:
+    cosi.io/provisioner: [1]
+  finalizers:
+    - cosi.io/finalizer [2]
+spec:
+  protocol: [3]
+  bucketPrefix: [4]
+  bucketClassName: [5]
+  bucketInstanceName: [6]
+status:
+  bucketAvailable: [7]
+```
+
+1. `labels`: added by COSI.  Key’s value should be the provisioner name. Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.
+1. `finalizers`: added by COSI to defer `BucketRequest` deletion until backend deletion succeeds.
+1. `protocol`: (required) specifies the desired protocol.  One of {“s3”, “gcs”, or “azureBlob”}.
+1. `bucketPrefix`: (optional) prefix prepended to a COSI-generated new bucket name, eg. “yosemite-photos-". If `bucketInstanceName` is supplied then `bucketPrefix` is ignored because the request is for access to an existing bucket.
```

Comment:
Is this prefix applied to the Bucket CR name? or the backend bucket name for greenfield case?

Reply:
No. The BR.name is up to the user -- it is their BR. The prefix is applied to greenfield backend bucket names.

Reply:
So if a user creates a greenfield BR without specifying a BR.spec.bucketInstanceName - what will be the B.name?

Reply:
BR.bucketInstanceName is the name of the cluster-scoped `Bucket` instance. If the user fills in this field then cosi knows we have brownfield (a new bucket will not be created). If B was created by the admin then its name is set by the admin. If cosi generated the B, then its name is non-deterministic; however, in this case the user who created the greenfield BR also creates a BAR that references their BR -- B's name is unimportant in this case.

Reply:
Got it, my only comment is that I would still prefer to define what is the generated B.name going to be rather than leave it unspecified.

Reply:
B.name, when generated by cosi, is defined in the kep as:
```
When created by COSI, the Bucket name is generated in this format: "<BR.namespace*>-<BR.name*>-". If an admin creates a Bucket, as is necessary for brownfield access, they can use any name. The uuid is unique within a cluster. * the first 10 characters of the namespace and the name are used.
```
Is this what you're looking for?

Reply:
@jeffvance thanks yes, I missed it.
I added a comment there (https://github.com/kubernetes/enhancements/pull/1383/files#r483814657).